Statistics Page

bb-froggy opened the pull request:
This PR adds a Statistics page to the Information sub-menu. It currently shows total games started, games won and lost.

I want to add more statistics from the list in #60, therefore this is still a draft PR. I opened it already to allow for feedback already.

#60 was assigned to @lauicibovi23, but I haven't seen any development for the last two years and https://github.com/lauicibovi23/FeudalTactics/tree/statistics_page is quite out-of-date, so I assumed I could tackle this. Possibly, @lauicibovi23 can have a look at my changes and provide feedback if you have any.

Disclaimer: I am not a regular Java developer, so if you find something odd, it is much more likely a beginner's mistake than some clever hack ;-).

## Open issues

- [x] Correctly count aborted games
- [x] Distinguish counts by some categories, most importantly AI level
- [x] Count map generations
- [x] Count which seed was played most
- [x] I used the preferences store to persist the statistics, but I've seen you could also create extra files with libGDX. From what I've read, the preferences store seems to be the right place, but prefer some confirmation.
- [x] There is an EditorScreen.java that publishes some of the same events as IngameScreen.java. Is there an editor in this game? I didn't know exactly how to handle it, but assumed it should not have any effect on the statistics. I am unsure whether passing the null parameter for the GameExitedEvent is the right way to do it.

Sesu8642 commented on core/src/main/java/de/sesu8642/feudaltactics/menu/statistics/ui/StatisticsEventHandler.java:
```diff
@@ -0,0 +1,67 @@
+    @Subscribe
+    public void handleGameStart(GameStartEvent event) {
+        statisticsDao.incrementGamesStarted();
+        statisticsSlide.refreshStatistics();
```
It is what I meant :)

Sesu8642 commented:
> There is an EditorScreen.java that publishes some of the same events as IngameScreen.java. Is there an editor in this game? I didn't know exactly how to handle it, but assumed it should not have any effect on the statistics. I am unsure whether passing the null parameter for the GameExitedEvent is the right way to do it.

The editor is unfinished. That's why it's not accessible by default. Your solution is fine. :)

If you want to play around with the editor, you can enable it in gameconfig.properties. :nerd_face:

bb-froggy commented:
I've incorporated your feedback. The seed played most has a heuristic implementation, as it currently stores only the number of times the last seed was played and the current record was played. It would be incorrect if you did something like the following:
- Play seed 42 for 3 times -> seed 32 would be record-holder with 3 plays
- Play seed 67 for 5 times -> seed 67 would be record-holder with 5 plays
- Play seed 42 for another 4 times -> seed 67 would (falsely) still be considered record holder with 5 plays, as this is more than 4 and the first three plays of 42 would not be considered, as they are not in a row

To measure the record holder seed with full accuracy, that would probably belong more to #50 .

bb-froggy commented:
BTW, copying the seed and then pasting into a new game doesn't work with the provided button on Android. I am not sure where the problem is. Pasting the seed via the virtual keyboard does work generally (but is awkward).

bb-froggy commented:
Ah, I understand now that copying and pasting is not just the seed, but needs a whole sharedPrefString. My current implementation stores how often a seed was played, but maybe what the user actually wanted is to know how often a particular set of game preferences was played?

In any case, thinking further about it, all information like how many games were won, lost, and aborted per type of difficulty, which is the preferences or seed played most, they all can be derived from the history of games (issue #50 ). I am now in doubt whether we should actually store this information redundantly for the statistics. One way to deal with it is to throw away this PR (though there are one or two mostly unrelated fixes in here that we should keep) and instead implement #50 properly.

The only thing that is not based on the history is the count of map generations, and that happens to be the only thing of this PR I have not yet implemented :-D.

Sesu8642 commented:
I agree that the more sophisticated statistics related to the seed should be derived from #50 and are out of scope for this PR. Your current implementation of the most played seed can be interesting but needs to be phrased well. Otherwise users might think it's buggy.

I think that this PR is a great base to build upon later and should be merged for sure. Once #50 is implemented, some of the stats could be refactored and be derived from the history.
It might make sense to save statistics redundantly anyway, as the history will need to have a maximum length for performance reasons and to limit disk space usage.

Sesu8642 commented:
Let's remove the most-played-seed for now then.
Absolutely feel free to work on the other issues as well! I'm very happy about your contributions :)

Sesu8642 commented:
@bb-froggy let me know when this is ready for a final review.

bb-froggy commented:
BTW, Thank you for this great game, @Sesu8642 !

Sesu8642 commented on core/src/main/java/de/sesu8642/feudaltactics/lib/gamestate/GameStateHelper.java:
```diff
@@ -119,6 +120,8 @@ public static void initializeMap(GameState gameState, List<Player> players, floa
         if (mapSeed == null) {
             mapSeed = System.currentTimeMillis();
+            // We don't send a SeedGeneratedEvent here, as this should be null only when this is called from the
+            // Editor, which is not really generating a new seed. Probably the value is just a dummy in this case?
```
There might be other use cases later. I think it's safe to assume that if the seed is null, no event needs to be sent. On a different note, I prefer sending the event from the UI classes (as you did), not here. That's why I'm thinking if this comment is helpful here or might be confusing.

bb-froggy replied:
I don't know the conventions for events in Java, Lombok, and this project, or wherever they are coming from. I just looked where other events were sent and which code paths "require" sending an event. I checked where currentTimeMillis was called, since this is what I considered a "seed generation" and by that definition, here is a seed generation without an event, hence the comment. But every "seed generation" that we want to appear in the statistics results from some action in the UI, so from that perspective, this piece of code doesn't require the event, so I guess it does more harm than help. I'll remove the comment.

Sesu8642 commented on core/src/main/java/de/sesu8642/feudaltactics/lib/gamestate/GameStateHelper.java:
```diff
@@ -46,6 +46,7 @@ public static GameState getCopy(GameState original) {
         if (original.getWinner() != null) {
             result.setWinner(copiedPlayers.get(original.getPlayers().indexOf(original.getWinner())));
         }
+        result.setBotIntelligence(original.getBotIntelligence());
```
Good catch! I wonder if this might have caused some issues before. But it seems like it was only used in the IngameScreen, EditorScreen and tests. All of those places seem to be uncritical.

Sesu8642 commented on core/src/main/java/de/sesu8642/feudaltactics/menu/play/ui/PlayMenuScreen.java:
```diff
@@ -72,6 +73,7 @@ private void initSandboxGame() {
         screenNavigationController.transitionToIngameScreen();
         eventBus.post(new NewGamePreferencesChangedEvent(savedPrefs));
         eventBus.post(new RegenerateMapEvent(savedPrefs.toGameParameters()));
+        eventBus.post(new SeedGeneratedEvent());
```
I assume you used this new event because the RegenerateMapEvent is sent too often? Maybe a comment can be added explaining the reason.

Sesu8642 replied:
Yes, that's a good place for it. Thanks!

Sesu8642 commented on core/src/main/java/de/sesu8642/feudaltactics/menu/statistics/StatisticsDao.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    public void incrementGamesWon(Intelligence aiDifficulty) {
+        incrementCountingStat(GAMES_WON_NAME);
```
Same for lost and aborted

Sesu8642 commented on core/src/main/java/de/sesu8642/feudaltactics/menu/statistics/ui/StatisticsEventHandler.java:
```diff
@@ -0,0 +1,64 @@
+/**
+ * Handles events for the statistics screen.
+ */
+public class StatisticsEventHandler {
```
Calling it StatisticsScreenEventHandler would be more consistent with the other handlers and the class description. But I think you chose this name because you also call the DAO from here.

In other cases (see PreferencesScreenEventHandler), I called the screen from the event handler and the screen then called the DAO. This way, the EventHandlers can stay very small. I think I prefer it that way.
But to be fair, I overused Events a bit and want to to simplify. So consider this comment a suggestion.

Sesu8642 reviewed (commented):
I left some smaller suggestions. Very good quality overall.

bb-froggy commented:
@Sesu8642 : Thank you for reviewing my PR and commenting it. I have addressed your comments and resolved those conversations where I was confident that the issue is now solved. For two of the comments, I have answered in the conversations and in one case changed some code, but I am not as confident that I cleared all your concerns, and therefore left the conversations unresolved. Please have another look.
